oauth2_http/java/com/google/auth/oauth2/IdTokenCredentials.java

```diff
@@ -0,0 +1,277 @@
+
+/**
+ * IdTokenCredentials provides a Google Issued OpenIdConnect token. <br>
+ * Use an ID token to access services that require presenting an ID token for
+ * authentication such as Cloud Functions or Cloud Run.<br>
+ * 
+ * The following Credential subclasses support IDTokens:
+ * ServiceAccountCredentials, ComputeEngineCredentials, ImpersonatedCredentials.
+ * 
+ * For more information see <br>
+ * Usage:<br>
+ * 
+ * <pre>
+ * String credPath = "/path/to/svc_account.json";
+ * String targetAudience = "https://example.com";
+ * 
+ * // For Application Default Credentials (as ServiceAccountCredentials)
+ * // export GOOGLE_APPLICATION_CREDENTIALS=/path/to/svc.json
+ * GoogleCredentials adcCreds = GoogleCredentials.getApplicationDefault();
+ * IdTokenCredentials tokenCredential = IdTokenCredentials.create(adcCreds, targetAudience);
+ * 
+ * // for ServiceAccountCredentials
+ * ServiceAccountCredentials saCreds = ServiceAccountCredentials.fromStream(new FileInputStream(credPath));
+ * saCreds = (ServiceAccountCredentials) saCreds.createScoped(Arrays.asList("https://www.googleapis.com/auth/iam"));
+ * IdTokenCredentials tokenCredential = IdTokenCredentials.create(saCreds, targetAudience);
+ * 
+ * // for ComputeEngineCredentials
+ * ComputeEngineCredentials caCreds = ComputeEngineCredentials.create();
+ * IdTokenCredentials tokenCredential = IdTokenCredentials.create(caCreds, targetAudience,
+ *     Arrays.asList(ComputeEngineCredentials.ID_TOKEN_FORMAT_FULL));
+ *
+ * // for ImpersonatedCredentials
+ * ImpersonatedCredentials imCreds = ImpersonatedCredentials.create(saCreds,
+ *     "impersonated-account@project.iam.gserviceaccount.com", null,
+ *     Arrays.asList("https://www.googleapis.com/auth/cloud-platform"), 300);
+ * IdTokenCredentials tokenCredential = IdTokenCredentials.create(imCreds, targetAudience,
+ *     Arrays.asList(ImpersonatedCredentials.INCLUDE_EMAIL));
+ * 
+ * // Use the IdTokenCredential in an authorized transport
+ * GenericUrl genericUrl = new GenericUrl("https://example.com");
+ * HttpCredentialsAdapter adapter = new HttpCredentialsAdapter(tokenCredential);
+ * HttpTransport transport = new NetHttpTransport();
+ * HttpRequest request = transport.createRequestFactory(adapter).buildGetRequest(genericUrl);
+ * HttpResponse response = request.execute();
+ *
+ * // Print the token, expiration and the audience
+ * System.out.println(tokenCredential.getIdToken().getTokenValue());
+ * System.out.println(tokenCredential.getIdToken().getJsonWebSignature().getPayload().getAudienceAsList());
+ * System.out.println(tokenCredential.getIdToken().getJsonWebSignature().getPayload().getExpirationTimeSeconds());
+ * </pre>
+ */
+public class IdTokenCredentials extends GoogleCredentials {
+
+  private static final long serialVersionUID = -2133257318957588431L;
+  private static final String CLOUD_PLATFORM_SCOPE = "https://www.googleapis.com/auth/cloud-platform";
+
+  private GoogleCredentials sourceCredentials;
```

Comment:
i'm marking this sequence as resolved as this squence is ultimately if 

```
public class IdToken extends AccessToken
```  
is done as described in https://github.com/googleapis/google-auth-library-java/pull/303#issuecomment-515244895

Reply:
Not all `GoogleCredentials` will work for refreshing the `IdToken`. We are not using any methods from `GoogleCredentials` so this should be an `IdTokenProvider`.

Reply:
Changing it will prevent a customer from using implict `GoogleCredential` which may be an underlying supported one.

that is, right now i can do something like this
```
export GOOGLE_APPLICATION_CREDENTIALS=/path/to/svc.json
```

then

```java
GoogleCredentials adcCreds = GoogleCredentials.getApplicationDefault();
IdTokenCredentials tokenCredential = IdTokenCredentials.create(adcCreds, targetAudience);
```

works

but if i use a userCredentials (which doens't implement it)

the follwing is thrown
```
IOException java.io.IOException: Provided sourceToken does not implement IdTokenProvider
```
--

I can do either but see its pretty convenient to allow both `GoogleCredentials` and explicit ones like `ComputeCredentials`, etc

Reply:
Since `IdTokenCredentials` is not a `GoogleCredentials` subclass, it doesn't really make a whole lot of sense to try to accept all `GoogleCredentials` types.

current:
```java
GoogleCredentials sourceCredentials = GoogleCredentials.getApplicationDefault();
IdTokenCredentials credentials = IdTokenCredentials.create(sourceCredentials, targetAudience);

// use the credentials
try {
  // fails when we try to use the credentials
  credentials.getRequestMetadata();
} catch (IOException ex) {
  if (ex.getMessage().equals("Provided sourceToken does not implement IdTokenProvider") {
    // wrong credentials type but nothing you can do about it
  } else {
    // handle other IOExceptions
  }
}
```

If we do allow `IdTokenCredentials.create(GoogleCredentials, String)`, we would want it to fail early with something like an `IllegalArgumentException` rather than waiting until it's used. 

with IllegalArgument:
```java
GoogleCredentials sourceCredentials = GoogleCredentials.getApplicationDefault();
IdTokenCredentials credentials;
try {
  credentials = IdTokenCredentials.create(sourceCredentials, targetAudience);
} catch (IllegalArgumentException e) {
  // test that IllegalArgumentException is because of invalid type and not some other possible illegal argument
}

// use the credentials
try {
  credentials.getRequestMetadata();
} catch (IOException ex) {
  // handle a request error or credentials error
}
```

I'm not a fan of this approach because the error handling is difficult. You'd rather have the user explicitly have to handle an unexpected credentials type if they are trying to use an IdTokenCredentials with an incompatible type.

change to IdTokenProvider:
```java
GoogleCredentials sourceCredentials = GoogleCredentials.getApplicationDefault();
if (!sourceCredentials instanceof IdTokenProvider) {
  // handle error message
}
IdTokenCredentials credentials = IdTokenCredentials.create((IdTokenProvider) sourceCredentials, targetAudience);

// use the credentials
try {
  credentials.getRequestMetadata();
} catch (IOException ex) {
  // handle a request error or credentials error
}
```

Reply:
fair enough; i'll keep `IdTokenProvider`